File: backend/app/permissions.py
"""Server-side capability matrix — the authoritative mirror of the frontend's
`lib/permissions.ts`. The frontend uses it for UI gating; the server enforces it.
"""

from __future__ import annotations

from app.models.enums import Role

Capability = str

_MATRIX: dict[Role, set[Capability]] = {
    Role.super_admin: {
        "approve-accounts",
        "create-users",
        "suspend-users",
        "upload-files",
        "assign-files",
        "view-all-schools",
        "view-own-school-teachers",
        "request-reports",
        "export-reports",
        "view-global-security",
        "view-school-security",
    },
    Role.school_admin: {
        # Monitoring-only.
        "view-own-school-teachers",
        "request-reports",
        "export-reports",
        "view-school-security",
    },
    Role.teacher: {
        "view-assigned-lessons",
        "use-ai-assistant",
    },
}


def can(role: Role, capability: Capability) -> bool:
    return capability in _MATRIX.get(role, set())

File: frontend/lib/permissions.ts
// Centralised role permissions. The frontend uses these for UI gating only.
// TODO: enforce the same matrix on the server.

import type { Role } from "@/types";

export type Capability =
  | "approve-accounts"
  | "create-users"
  | "suspend-users"
  | "upload-files"
  | "assign-files"
  | "view-all-schools"
  | "view-own-school-teachers"
  | "request-reports"
  | "export-reports"
  | "view-global-security"
  | "view-school-security"
  | "view-assigned-lessons"
  | "use-ai-assistant";

const matrix: Record<Role, Capability[]> = {
  "super-admin": [
    "approve-accounts",
    "create-users",
    "suspend-users",
    "upload-files",
    "assign-files",
    "view-all-schools",
    "view-own-school-teachers",
    "request-reports",
    "export-reports",
    "view-global-security",
    "view-school-security",
  ],
  "school-admin": [
    // Monitoring-only. No approvals, no uploads, no user creation, no assignments.
    "view-own-school-teachers",
    "request-reports",
    "export-reports",
    "view-school-security",
  ],
  teacher: ["view-assigned-lessons", "use-ai-assistant"],
};

export function can(role: Role, cap: Capability): boolean {
  return matrix[role].includes(cap);
}
